R/tspa.R

```diff
@@ -58,222 +111,205 @@
 #'      group = "school")
-#' }
 
 
+tspa <- function(model, data, reliability = NULL, se = NULL,
+                 vc = NULL, cross_loadings = NULL, ...) {
+  if (!is.null(reliability)) {
+    stop("tspa() currently does not support reliability model")
+  }
+  if (!is.data.frame(se)) {
+    se <- as.data.frame(as.list(se))
+  }
+  multigroup <- nrow(se) == 1 | is.list(vc)
 
+  if (multigroup) {
+    if (is.null(vc)) { # SE
+      tspaModel <- tspaMultipleGroupSe(model, data, se)
+    } else { # covariance
+      tspaModel <- tspaMultipleGroupMF(model, data, vc, cross_loadings)
+      data <- do.call(rbind, data)
+    }
+  } else {
+    if (is.null(vc)) { # SE
+      tspaModel <- tspaSingleGroup(model, data, se)
+    } else { # covariance
+      tspaModel <- tspaSingleGroupMF(model, data, vc, cross_loadings)
+    }
+  }
 
-# # modify for multiple group tspa model
-# tspa <- function(model, data, reliability = NULL, se = NULL) {
-#   var <- names(reliability)
-#   len <- length(reliability)
-#
-#   if (len < 2) {
-#     stop("Reliability len is smaller than 2, unable to build model. Reliability needs to consist of at least 2 variables.");
-#   }
-#   col <- colnames(data)
-#   fs <- paste0("fs_", var)
-#
-#   if (length(fs) < 2) {
-#     stop("Data column len is smaller than 2, unable to build model. Data needs to consist of at least 2 columns.");
-#   }
-#
-#   tspaModel <- rep(NA, len)
-#   latent_var <- rep(NA, len)
-#   error_constraint <- rep(NA, len)
-#   latent_variance <- rep(NA, len)
-#   reliability_constraint <- rep(NA, len)
-#
-#   if(fs %in% col == FALSE) {
-#     stop("data columns does not match with reliability")
-#   }
-#
-#   for (x in 1:len) {
-#     latent_var[x] <- paste0(var[x], " =~ 1 * ", fs[x], "\n")
-#     error_constraint[x] <- paste0(fs[x], " ~~ ev", x, " * ", fs[x], "\n")
-#     latent_variance[x] <- paste0(var[x], " ~~ v", x, " * ", var[x], "\n")
-#     reliability_constraint[x] <- paste0("v", x, " == ", toString(reliability[x]), " / ", toString(1 - reliability[x]), " * ev", x, "\n")
-#   }
-#
-#   latent_var_str <- paste(latent_var, collapse="")
-#   error_constraint_str <- paste(error_constraint, collapse="")
-#   latent_variance_str <- paste(latent_variance, collapse="")
-#   reliability_constraint_str <- paste(reliability_constraint, collapse="")
-#   tspaModel <- paste0(latent_var_str, error_constraint_str, latent_variance_str, model, "\n", reliability_constraint_str)
-#
-#   tspa_fit <- sem(model = tspaModel,
-#                   data  = data)
-#
-#   # tspaModel <- cat(tspaModel)
-#   attributes(tspa_fit)$tspaModel <- tspaModel
-#   # to access the attribute, use attr(x,"tspaModel")
-#   return (tspa_fit)
-# }
+  tspa_fit <- sem(model = tspaModel,
+                  data  = data,
+                  ...)
+  # to access the attribute, use attr(x,"tspaModel")
+  attr(tspa_fit, "tspaModel") <- tspaModel
+  return(tspa_fit)
+}
 
+tspaSingleGroup <- function(model, data, se = NULL) {
```

Comment:
Do we have a naming convention for functions in our package?

Reply:
I don't think we have, but I think it's good to have one. We may want to add alias to the functions first so that it doesn't break the old code.